src/frontend/src/optimizer/plan_node/generic/agg.rs

```diff
@@ -39,6 +39,123 @@ use crate::utils::{
+                | AggKind::BoolOr
+                | AggKind::JsonbAgg
+                | AggKind::JsonbObjectAgg
+                | AggKind::PercentileCont
+                | AggKind::PercentileDisc
+                | AggKind::Mode
+        };
+    }
+    pub(crate) use unimplemented_in_stream;
+
+    /// [`AggKind`]s that should've been rewritten to other kinds. These kinds should not appear
+    /// when generating physical plan nodes.
+    macro_rules! rewritten {
+        () => {
+            AggKind::Avg
+                | AggKind::StddevPop
+                | AggKind::StddevSamp
+                | AggKind::VarPop
+                | AggKind::VarSamp
+        };
+    }
+    pub(crate) use rewritten;
+
+    /// [`AggKind`]s of which the aggregate results are not affected by the user given ORDER BY
+    /// clause.
+    macro_rules! result_unaffected_by_order_by {
+        () => {
+            AggKind::BitAnd
+                | AggKind::BitOr
+                | AggKind::BitXor // XOR is commutative and associative
+                | AggKind::BoolAnd
+                | AggKind::BoolOr
+                | AggKind::Min
+                | AggKind::Max
+                | AggKind::Sum
+                | AggKind::Sum0
+                | AggKind::Count
+                | AggKind::Avg
+                | AggKind::ApproxCountDistinct
+                | AggKind::VarPop
+                | AggKind::VarSamp
+                | AggKind::StddevPop
+                | AggKind::StddevSamp
+        };
+    }
+    pub(crate) use result_unaffected_by_order_by;
+
+    /// [`AggKind`]s that must be called with ORDER BY clause. These are slightly different from
+    /// variants not in [`result_unaffected_by_order_by`], in that variants returned by this macro
+    /// should be banned while the others should just be warned.
+    macro_rules! must_have_order_by {
+        () => {
+            AggKind::FirstValue
+                | AggKind::LastValue
+                | AggKind::PercentileCont
+                | AggKind::PercentileDisc
+                | AggKind::Mode
+        };
+    }
+    pub(crate) use must_have_order_by;
+
+    /// [`AggKind`]s of which the aggregate results are not affected by the user given DISTINCT
+    /// keyword.
+    macro_rules! result_unaffected_by_distinct {
+        () => {
+            AggKind::BitAnd
+                | AggKind::BitOr
+                | AggKind::BoolAnd
+                | AggKind::BoolOr
+                | AggKind::Min
+                | AggKind::Max
+                | AggKind::ApproxCountDistinct
+        };
+    }
+    pub(crate) use result_unaffected_by_distinct;
+
+    /// [`AggKind`]s that are simply cannot 2-phased.
+    macro_rules! simply_cannot_two_phase {
+        () => {
+            AggKind::ArrayAgg
+                | AggKind::JsonbAgg
+                | AggKind::JsonbObjectAgg
+                | AggKind::PercentileCont
+                | AggKind::PercentileDisc
+                | AggKind::Mode
```

Comment:
Shouldn't approx count distinct fall under this category? 🤔

Reply:
And `StringAgg` too?

Reply:
Seems to be the source of errors.

Reply:
I guess the `src/meta/src/stream/stream_graph/fragment.rs:279:9` is because `visit_stream_node_tables_inner` didn't correctly visit state table of `DistinctApproxCount`. While OTOH 2-phase `sum0(hyperloglog)` may indeed not equal to `hyperloglog`, not quite sure. I'll add it to this category.

For `StringAgg`, I was thinking that when user doesn't provide `ORDER BY`, the call can actually be 2-phased by just concatenating strings in arbitrary nondeterministic order.

Reply:
> For StringAgg, I was thinking that when user doesn't provide ORDER BY, the call can actually be 2-phased by just concatenating strings in arbitrary nondeterministic order.

Agree, I suppose there's some other bug around it then. Maybe it is executing in non-append-only. We should add another check for that.

An e2e test will be great to test two-phase string_agg as well, I don't think it is currently tested. Same goes for any other two-phase agg we have just started supporting in this PR.

Reply:
Seems 2-phase StringAgg has other problem, that is the final phase can't get `delim` argument from input. Maybe better to disable it again for now😇.